### ISSUE TITLE
Not all the live followed channels show up

### ISSUE THREAD

**Issue by mathkkad** [bhb27/smarttv-twitch#4]
Hi,

Thank you for all the hard work in making this wonderful app!

I have the Samsung TV: 50KU6072  and i loaded the app according to the instructions from my Windows 10 64bits Version 1703 (OS Build 15063.726) and i have the following problem:

Not all my followed channels from my twitch user name show up. I can see only 9 (can vary, sometimes i get 13 out of 30+ live followed channels) although i have over 30 followed channels live (seen on my PC web browser, twitch.tv site)

Also is there a way to see VODs on this app? And if not, is such a feature planned?

Thank you

**Comment by bhb27** [comment]
Are you shore they are all live streaming at the moment?
the app only show active live streams on fallowed channels the same as the bellow twitch link
https://www.twitch.tv/directory/following/live

I made a test account fallow a lot of channels that are live right now up to that link showed me 36 live stream from the channels I fallow, then on the app I get the same result all 36 was there.

So I ask are you shore?
how many games you fallow?
I doubt that it can be but I can test here, I only fallowed 9 games on that test account...

**Comment by mathkkad** [comment]
Hi @bhb27 and thank you for your reply,

Just to confirm i am a 35 years old person and i have experience with PCs since i was 7 years old (i know what i am talking about - so you can be sure about it)

I have been watching twitch.tv for a couple of years now and i have 112 followed channels as follows at the moment of my posting:

22 live followed channels
4 vodcast channels
86 offline channels (some may be hosting someone - hosted channels)
(data confirmed by twitch.tv - on chrome browser and also on Twitch Desktop app - PC software)

My TV shows:
12 live followed channels
30 hosted channels

For example a live followed channel last night was available on TV last night and today the same live channel is not - seems the TV selection of live channels from my account is random based of starting the app, closing/opening the TV.

If you need any further info or pictures (even short video of this if that can be posted here) feel free to ask, as i want to assist you in any way i can.

Once again thank you for you interest and for this wonderful app. Hope this issue is sorted up.

Best Regards

**Comment by bhb27** [comment]
I see what the cause of the problem is now, the more detailed description helped...

The app is not design to see over 99 user followed channels, because twitch doesn't allows request above 100 at a time...

So the result you was seeing was a little random because it was only showing you the live channels from the last 99 channels that you sub...

The proper fix is hard work I have no time for this week, maybe next weekend I have a small break and proper fix the problem...

On the mean time I set the app on the above commit to do that 99 request base on the last channel a user fallows shorted broadcast live time, so the one that broadcast last will be first on the list with theoretically will make the result to show the proper list for you as long there ins't 99+ channels live and the twitch result is properly up to date regarding the times...

When I have time I will make the proper check for 99 and + wherever the total result for fallowed is for it user...


please test the above commit and let me know the result...
:+1:

**Comment by bhb27** [comment]
I'm not able to see any VOD live stream from the channel I fallow...

So i did the above to see if the "Live channels User" show the VODs with the normal stream for you if it does, you let me know with info from the application log...

to see the log info you need to run the app as debug:

before you need to make shore [this line](https://github.com/bhb27/smarttv-twitch/commit/1c7f75361f1b925a2680bc1e1f18de264f41b866#diff-4cda972c55f6fc61bd4954d7c1c734ecR503) doesn't start have those "//" then...

instead of "right click > Run as > 1 Tizen wen Application", click on the top left corner "ladybug
" icon, and install the app as debug right after the app finish installing will pop a new web browser window in the computer, on that will be the debug information...

on the app click OK, then wait the pop window load in the PC, on the top bar click on console it will be probably empty, then just go to User after setting yours user and see the output of the log, I need you to share that output as a text here on a comment here

**Comment by mathkkad** [comment]
Hi,

You have identified well the problem, now it works!

I have installed the latest version of the app and i can confirm that i see now all my live followed channels (including the vodcasts - people that are not live but are playing live a previous broadcast). But compared to the other version when i exit the TV app now it asks for my Twitch User ID  every time, i remember the previous version remembered my Twitch User ID until i changed it manually

After your last post i have also run (installed) the app with the ladybug icon with:

 //console.log(JSON.stringify(response));       line is with "//" in front as you said,

When i install the app i get an error:

![installation error promt](https://user-images.githubusercontent.com/34389600/33800854-5840d5aa-dd52-11e7-862f-1908770670fe.jpg)

The console shows the following info:

![debug console](https://user-images.githubusercontent.com/34389600/33800856-64719c74-dd52-11e7-8883-f96933133d71.jpg)

Here are my followed channels:

![1](https://user-images.githubusercontent.com/34389600/33800859-7ddfbe5c-dd52-11e7-9e02-370159ebf7d1.png)

![2](https://user-images.githubusercontent.com/34389600/33800862-835a860a-dd52-11e7-806b-9a69f389f60c.png)

![3](https://user-images.githubusercontent.com/34389600/33800865-88bf46da-dd52-11e7-8c32-0c6d13c594b7.png)

![4](https://user-images.githubusercontent.com/34389600/33800869-8d46663e-dd52-11e7-9e08-58e2b876a7b0.png)

![5](https://user-images.githubusercontent.com/34389600/33800872-933792a2-dd52-11e7-902a-24eeca2a6c03.png)

![6](https://user-images.githubusercontent.com/34389600/33800874-97ef5b68-dd52-11e7-95f2-99b7d1fb183f.png)

But as i said, now i can see all the followed live channels.

Is there a way to see a channel`s VOD (video on demand = the previous broadcasts of a channel that are now as offline videos?)

Please tell me if you need more info and thank you for your fast patching of the problem

**Comment by bhb27** [comment]
I manage to get all I needed, even even though the logs did not output as expected...

Those tizen error are from problems in platform it self, not related to a app error and they don't cause app error, it can be dismissed...
the console one are normal related to http requests that fail, just internet related that is expected the app handle it all and retry...

I add a extra info to the streamer name to differentiate live that are lives from VOD, recheck everything and seems to me that is all fine.

If you notice a problem related to this let me know.

When I finish properly coding the app to support over 99 fallowed channel I will inform here so you can properly test.

The offline VOD is a challenge as is a separated way to get and display all the related information, and playing a offline VOD is not the same as playing a live stream is diferente with makes it more complicated, so basically need to design a "separated app" in the app to deal with it all and that is days and days of work something that I don't have, but I will probably do just like the smarthub implementation, that toke me a few days to make and I did little by little on a separated branch, but I have no idea when I will start, probably next January I have started and finished it or at least a beta version that supports it...

:+1:

**Comment by mathkkad** [comment]
I am glad i could be of assistance and you gathered all the data needed @bhb27.
Yes, now it is working fine, your idea to let the app see the last 99 live channels was excellent

It would be amazing to have a complete app and also have offline VODs included even it will be in a month or 2

So, in conclusion:
1. If i notice a problem related to this i will inform you
2, Could you please make the app save the twitch user name (as i remember it was in the previous version) because it is very annoying to always enter my twitch user name every time i open the tv (and sometimes in the same session if i switch between remote control buttons C, A and B - that asks again twitch user name) plus it makes the option "press Extra + B to change user name" useless - inform me to test it
3. When you notify me of the official support of the over 99 channel i will test it and give you all the feedback data
4. Please announce me when you have a version with offline VODs included so i can also test that and give you feedback

Thank you for your help and dedication

**Comment by bhb27** [comment]
@mathkkad 
2. https://github.com/bhb27/smarttv-twitch/commit/a6d5d95f1c2620bf02dac722d31edcd4e46ee0f4

**Comment by mathkkad** [comment]
Yes, you are right @bhb27, i did not saw that.
I have installed the app now

I can confirm that the user name remains saved now if you close/open the app.

But like in the previous version (that the user name remains saved) when i press "Extra + B" the twitch user name appears but i cannot delete it with the TV "included virtual keyboard" using the TV remote, i can only add letters/numers/characters to it but i cannot "delete all" or "backspace" to delete. I can only delete it if i enter in the TV a phyiscal usb keyboard. Only that works.

It is no major problem to me but this is my feedback

**Comment by bhb27** [comment]
I try to fix that before I miss one line prevent the fix
know after the below is 100% fixed
https://github.com/bhb27/smarttv-twitch/commit/f078e25d6ace27ee7fa99c208f05ee30bd238480

:+1:

**Comment by mathkkad** [comment]
Ok, i have downloaded the app again from "Code" -> "Clone or Download" and tested it, finding out the following:

1. Pressing down "channel down" on the TV remote no longer refreshes the current section but moves you to the previous (example: being in "B - User section" and pressing "channel down" no longer refreshes "B" but switches to "A - live streams")
2. Pressing B takes you to entering new twitch user (previous "Extra+B") and the only way to access "B" with your saved user is to load "A" than press "channel up on the remote" to get to "B section" that has your twitch user name saved
3. I can no longer access chat feed while a stream is loaded on the TV by pressing "channel up" remote TV button (this also happened in the previous app version where i could not delete the twitch user name)

For me these are minor problems that can be passed by but i am telling you just to help with the project feedback if you need it and it is ok with you

**Comment by bhb27** [comment]
Yes I know, I update the readme when I made that change, you can see that the app top bar has a new function on CH press key...

https://github.com/bhb27/smarttv-twitch#live-screen
https://github.com/bhb27/smarttv-twitch#playing-a-video-screen

**Comment by mathkkad** [comment]
Ok, i see

I can confirm it works fine that way.

For me is the GUIDE button since i have another kind of remote for my TV. Also good to know for all who have this kind of remote.

![samsung remote](https://user-images.githubusercontent.com/34389600/33853500-da00b39c-dec6-11e7-993a-a5dc58e674bb.jpg)

GUIDE - open/close chat when a stream is open (CH up with chat open to reposition)
GUIDE - refresh current section (CH up/down move between sections A,B,C,D)

**Comment by bhb27** [comment]
Yes that helps a lot, I update the readme with extra info...
btw I have a samsung remote from a home theater, that works well to control the TV but it doesn't have a **Guide** key, so I re-map the **info** key to do on the app the same function of the guide one just in case someone need it...

:+1:

**Comment by mathkkad** [comment]
Glad that i could help.

Sounds good, features for everyone no matter the remote!

Looking forward to new updates to the project and testing them up

**Comment by bhb27** [comment]
I made a beta version of the new user interface that theoretically fully support any number of fallowing channels...

https://github.com/bhb27/smarttv-twitch/tree/user_rework

All changes are on above branch...
**If you open one of the user options LIVE/HOSTS/GAMES change bettwin them use channel key...**

If you find anything not working let me know...
On the weekend I finish resolving any bug that we can find :+1:

**Comment by bhb27** [comment]
@mathkkad  just in case you tested before this post, I fix some bugs related to loading all user live fallowing that wasn't loading it all some times...

Plus I add multi user support...

So if you test before this post and found a bug related to not see all user live fallowing, try again if you find any bug with the new implementation please let me know

:+1:

**Comment by bhb27** [comment]
VOD support is up on the bellow branch
https://github.com/bhb27/smarttv-twitch/tree/user_rework


Is wip can be improve and may have minor bugs...

**Comment by mathkkad** [comment]
Hi,
Sorry for the delay but i thought it will take more time till you make an update

I have tested 40min ago the latest build from:
https://github.com/bhb27/smarttv-twitch/tree/user_rework

Here are my findings:

1. The white contour over the selected channel helps better identify the selected channel but the white border over the name / title/ user number of the channel makes things **less** visible / readable than before. The classical one where you just see them as they are native would be better or maybe change selection color for better visibility

2. The channels load **much** faster now. I also use the twitch app from samsung tv store and when i switch between them the time till the image appears is 2 secs for the twitch store one and 4 seconds for your app down from aprox 8 seconds which is a great improvement. 

3. The INFO key works the same as the GUIDE key, all is identical and coded well

4. I CANNOT add my user name to the B user section. After i enter my user name by tv remote i cannot press done. I only managed to press ok by an external usb keyboard

5. I entered my user name misspelled once and i could not enter my user name again as i could no more edit the user name field unless exit the app first. After i exit and re enter the app i entered my user name and pressed ok (by usb keyboard as it does not work by tv remote) and although i entered my correct user name and press enter the app still uses the misspelled one. I had to reinstall the app to add my correct user name

6. I love the way now everything is separated into categories: live channels/ live hosts... so on

7. The offline VOD seemed to work just fine. Maybe a slider that can also be controlled by left/right arrows would be nice, but still is good as it is. Also 5 minutes back/rewind increments would also be welcomed.

8. I love that if i press 10 mins forwards (or any time increment) it remains selected so i can press it again without re selecting it so i can faster adjust the video to the position i want

9. One offline VOD problem i encountered: when starting a offline VOD let say it is at 00:00:08 seconds when i press 10 seconds back or any number of time back so i can get the video again to 00:00:00 (video start) it is still at 00:00:08 or similar and if i press it more times with various back time increments (2 mins back, 10 mins back) it freezes the video and it cannot be started by any option except exit it and reload the video again.

I will test it further and give you feedback to what i find. Please keep me up to date to your development so i can test more

**Comment by mathkkad** [comment]
I have updated the above post a couple of times with more info.

I have also discovered a offline VOD problem:

I only get in past VOD section only 16x6 rows = 96 channels. This means i do not get all my followed channels. As they are sorted alphabetically i only get till letter L. This means i do not get offline VOD channels like "Zotbot", "Prismat" or "xMinks". Is there a limit in place?

Also all offline VOD channels i tried work great except "GCWC" where the streams remains freezed with "loading" message and i cannot "back" i can only "exit" app for it to stop and have to re enter app (the browser/ twitch desktop offline VOD for "GCWC" work fine)

**Comment by bhb27** [comment]
> I CANNOT add my user name to the B user section. After i enter my user name by tv remote i cannot press done. I only managed to press ok by an external usb keyboard

I don't understand the above as I don't have that problem, are you using the on screen keyboard?
because that is the only one I use besides a usb that I don't use most of the time and that is not the problem, the user name may help as I can input there and try...

> I entered my user name misspelled once and i could not enter my user name again as i could no more edit the user name field unless exit the app first. After i exit and re enter the app i entered my user name and pressed ok (by usb keyboard as it does not work by tv remote) and although i entered my correct user name and press enter the app still uses the misspelled one. I had to reinstall the app to add my correct user name

the above is fixed, but please try...

> Also 5 and 15 minutes back/rewind increments would also be welcomed.

I add some more times

> One offline VOD problem i encountered: when starting a offline VOD let say it is at 00:00:08 seconds when i press 10 seconds back or any number of time back so i can get the video again to 00:00:00 (video start) it is still at 00:00:08 or similar and if i press it more times with various back time increments (2 mins back, 10 mins back) it freezes the video and it cannot be started by any option except exit it and reload the video again.

OK this is a TV related, as my doesn't do that, it goes to 0 if let's say current time it 1min and I do a jump back to 2min or something like... But there is a way to work around that so I add a function that will seek to 0 when time to jump is lower then "0"

> A Offline VOD channel (maybe even for live channels) search tool in the future would be helpful as it may be hard to find the channel that you want to see the offline VOD from if you have hundreds of followed channels

I set it to alphabetical order as it makes easier to find, the open function "D" can open a live channel by its name, I can mod it with time so it doesn't open the live channel, instead show if is live with a option to open it's VOD...


> I only get in past VOD section only 16x6 rows = 96 channels. This means i do not get all my followed channels. As they are sorted alphabetically i only get till letter L. This means i do not get offline VOD channels like "Zotbot", "Prismat" or "xMinks". Is there a limit in place?

The app wasn't calculating that you was going down, it starts with 96 as you go down it will load more if you have more to load, I fix the calculation and now is loading all OK...

> Also all offline VOD channels i tried work great except "GCWC" where the channel remains frozen with "loading" message and i cannot "back" i can only "exit" app for it to stop and have to re enter app (the browser/ twitch desktop offline VOD for "GCWC" works fine)

I manage to get this on time, as the problem here was a 404 video the below
https://www.twitch.tv/videos/205849380
twitch fix the video so it doesn't show in the list as you can see if you click it will not play, but before twitch remove it the app try to load the video info with causes the app to freezes on a 404 loop, I add a function that will prevent that from happening and will show the video with a interrogation thumbnail and a  info about 404 status, and will not allow the user to click open on it, this will prevent this bug from freezing the app.


So please check again with time anything that I may have miss from your post and also those that I fix just in case is not 100% fixed...

**Comment by mathkkad** [comment]
@bhb27 i have tested the app again

> "I CANNOT add my user name to the B user section. After i enter my user name by tv remote i cannot press done. I only managed to press ok by an external usb keyboard"

This means if i go to the "B" "User Section" a field appears so you can enter the user name. In previous versions when the field was empty the option on the TV on screen keyboard controlled by the TV remote of the "Enter key"/"Done" button it was grayed out naturally as the field is empty and no user name is there to enter. As you filled in a user name the buttons on the TV on screen keyboard controlled by the TV remote "Enter key"/"Done" became white so you can press them to input the user name.

In this version even if you fill in a user name the buttons are grayed out "Enter key"/"Done" meaning the app sees what you imputed as non existent = the app thinks the user name input field is blank although you have entered one.

(See picture below)

![20171218_201608](https://user-images.githubusercontent.com/34389600/34122142-1bcc85c2-e434-11e7-8a57-315631a7f395.jpg)


The only way that i could enter a user name was by an external usb keyboard

I have tested the other issues and are now solved:

> "Also 5 and 15 minutes back/rewind increments would also be welcomed"

DONE

> "5.	One offline VOD problem i encountered: when starting a offline VOD let say it is at 00:00:08 seconds when i press 10 seconds back or any number of time back so i can get the video again to 00:00:00 (video start) it is still at 00:00:08 or similar and if i press it more times with various back time increments (2 mins back, 10 mins back) it freezes the video and it cannot be started by any option except exit it and reload the video again."

It now goes to 00:00:00. Although adding time over the video max length does nothing and does not say why (it is ok for me but somewhat unexpected behavior, maybe a message to notify the user that you cannot go that far in the video future is a thing to consider)

> "6.	Twitch channels appear as they are in the desktop app / browser (regarding the 99 limit) - that is fine (i will also see if it is the same tonight when more channels are live)"

I can confirm all is fine all videos are the same as the twitch desktop app / browser

> "I have also discovered a offline VOD problem:
I only get in past VOD section only 16x6 rows = 96 channels. This means i do not get all my followed channels. As they are sorted alphabetically i only get till letter L. This means i do not get offline VOD channels like "Zotbot", "Prismat" or "xMinks". Is there a limit in place?"

All is fine, i can now see all my followed channels. I will test more of them to see if i find any channel with offline VODs not fuctioning

> "Also all offline VOD channels i tried work great except "GCWC" where the channel remains frozen with "loading" message and i cannot "back" i can only "exit" app for it to stop and have to re enter app (the browser/ twitch desktop offline VOD for "GCWC" works fine)"

"GCWC" now works fine

**Comment by bhb27** [comment]
> This means if i go to the "B" "User Section" a field appears so you can enter the user name. In previous versions when the field was empty the option on the TV on screen keyboard controlled by the TV remote of the "Enter key"/"Done" button it was grayed out naturally as the field is empty and no user name is there to enter. As you filled in a user name the buttons on the TV on screen keyboard controlled by the TV remote "Enter key"/"Done" became white so you can press them to input the user name.
> 
> In this version even if you fill in a user name the buttons are grayed out "Enter key"/"Done" meaning the app sees what you imputed as non existent = the app thinks the user name input field is blank although you have entered one.
> 
> (See picture below)
> 
> ![20171218_201608](https://user-images.githubusercontent.com/34389600/34122142-1bcc85c2-e434-11e7-8a57-315631a7f395.jpg)
> 
> If i press the TV remote RETURN button it erases the imputed user name
> 
> The only way that i could enter a user name was by an external usb keyboard


And all that start on this version of the APP? or was always like this.

because like I wrote before here is fine "Done" is always there even if I do not enter anything, noting on the app can change that behavior, I did not change the code related to that also is the same as always was, the others things around yes are different but they don't affect.

So can you try to fix on yours TV? reboot it power off and on from the wall.
change keyboard config, there is a gear icon on yours ss, maybe is because you have a keyboard connected if it is connected, removing problem solve I don't know... doesn't seems a app problema at all more a TV bug.

**Comment by mathkkad** [comment]
> And all that start on this version of the APP? or was always like this.

because like I wrote before here is fine "Done" is always there even if I do not enter anything, noting on the app can change that behavior, I did not change the code related to that also is the same as always was, the others things around yes are different but they don't affect.

So can you try to fix on yours TV? reboot it power off and on from the wall.
change keyboard config, there is a gear icon on yours ss, maybe is because you have a keyboard connected if it is connected, removing problem solve I don't know... doesn't seems a app problema at all more a TV bug.

SOLVED @bhb27 - rebooted the TV and now it works fine. In previous app versions all was fine, problem was only in this version. I don`t have a usb keyboard attached to the tv. I only attached it to bypass this problem.

One more thing:

Regarding "Past Broadcast" section channel list: if let`s say i open channel 45 and than return to "Past Broadcast" section channel list it refreshes and goes to channel 1 and i have to go again to channel 45. Can you make it to remember the 45 (or last watched channel position) like it is for "Live channels" where you see a channel and if you return to the list it saves the last position and if you want to refresh you can do it manually? Because it is annoying to go over and over 45, 75 or more channels to be again where you were

**Comment by bhb27** [comment]
> SOLVED @bhb27 - rebooted the TV and now it works fine. In previous app versions all was fine, problem was only in this version. I don`t have a usb keyboard attached to the tv. I only attached it to bypass this problem.

I check samsung keyboard implementation, there is a call that was not be using it may help to prevent those odd behaviors if they was generated via this app, plus I manage to map the backspace and delete all on screen keyboard keys with is a plus...

> Regarding "Past Broadcast" section channel list: if let`s say i open channel 45 and than return to "Past Broadcast" section channel list it refreshes and goes to channel 1 and i have to go again to channel 45. Can you make it to remember the 45 (or last watched channel position) like it is for "Live channels" where you see a channel and if you return to the list it saves the last position and if you want to refresh you can do it manually? Because it is annoying to go over and over 45, 75 or more channels to be again where you were

When you open a new page that contains thumbnails the previously need to be empty as there is only one main thumbnails page, so you need to reload it all the time, is possible to duplicate if is not too much work I make this week need to check the implementation necessary and the behavior...

> And is there a way to pause a offline VOD? i did not managed to find one

that function is already there even to live streams but not to yours remote...
I only test on the "smart remote" it has a dedicated "media play" key the app wasn't mapping the single play or pause key, so I map those keys not as a single function for it, one mono function, if is playing click one (play or pause) will pause, if i pause click one (play or pause) it will play, those key can also be used as the enter key, to open a stream for example... I test on my backup remote it has single play pause it is working here...

:+1:

**Comment by mathkkad** [comment]
> When you open a new page that contains thumbnails the previously need to be empty as there is only one main thumbnails page, so you need to reload it all the time, is possible to duplicate if is not too much work I make this week need to check the implementation necessary and the behavior...

Ok, thank you for that, there is no hurry. Would be useful if it does not too much time

Unfortunately no luck here with pause function, i have tried in both live streams and offline VOD all the remote keys a couple of times (including the media keys and everything). No one pauses the stream. Could it be mapped to something else like EXTRA button or PRE-CH or something?

**Comment by bhb27** [comment]
you need to update the app, I only remap the keys after you told me it doesn't work.

all the function was already there, but the keys not, so update and retest.

**Comment by mathkkad** [comment]
I can confirm now it does work. Both "play" and "pause" media keys work

Thank you for that

**Comment by bhb27** [comment]
I will close this if you have a bug please open a new issue this is too big and I don't wanna to get lost...

I delete the branch https://github.com/bhb27/smarttv-twitch/tree/user_rework
and add all it's change to the main branch https://github.com/bhb27/smarttv-twitch/tree/tizen

I add a featuring to go back to channel that was open in past broadcast...

But I did not like the performance, because of that I will start the app from scrash, I'm working on a the branch https://github.com/bhb27/smarttv-twitch/tree/new

The day I finish I will update the read.me to incude a new gif and will change the title with a version tag info, you can try that if you wanna when I don't finish, the new branch can be installed without removing the tizen branch one, just create a new project on sdk and use that source, but don't open issues regarding it until I released it's change to the tizen branch, as it is a wip project will take a few weekes to finish but the result will be way better in relation to current app performance, looks, etc related to code, debugging and add features will be really improved..

:+1:
